### PR TITLE
Add reload capability for Vault listener certs

### PR DESCRIPTION
**Pull request by jefferai**
This can be performed at any time, but must be performed by an administrator, and as an added safety check, it currently will only re-read the same files initially given in the config at startup time. We can consider changing this later if we want (and the mechanism for how this would happen is already plumbed through).

This also generally adds a mechanism for triggering a reload; the core can call the provided reload functions with the data passed in via the API. It's up to the reload functions to do the right thing, but I don't anticipate that most of them will be very complicated. Most things will still require restart Vault.

Still to do:

* Docs
* CLI command/API

### CODE REVIEW

File: command/server.go [hunk 1]
```diff
@@ -333,11 +343,23 @@ func (c *ServerCommand) Run(args []string) int {
 	logGate.Flush()
 
 	// Wait for shutdown
-	select {
-	case <-c.ShutdownCh:
-		c.Ui.Output("==> Vault shutdown triggered")
-		if err := core.Shutdown(); err != nil {
-			c.Ui.Error(fmt.Sprintf("Error with core shutdown: %s", err))
+	shutdownTriggered := false
+	for {
```
Comment (armon) [inline command/server.go]: May as well do `for !shutdownTriggered {` and remove the if block

Comment (jefferai) [inline command/server.go]: Hah, I've gotten so used to the infinite-loop-using-for-to-keep-selecting paradigm that it didn't even cross my mind.

File: command/server/listener.go [hunk 2]
```diff
@@ -73,15 +77,54 @@ func listenerWrapTLS(
...
 	if !ok {
-		return nil, nil, fmt.Errorf("'tls_min_version' value %s not supported, please specify one of [tls10,tls11,tls12]", tlsvers)
+		return nil, nil, nil, fmt.Errorf("'tls_min_version' value %s not supported, please specify one of [tls10,tls11,tls12]", tlsvers)
 	}
 	tlsConf.ClientAuth = tls.RequestClientCert
 
 	ln = tls.NewListener(ln, tlsConf)
 	props["tls"] = "enabled"
-	return ln, props, nil
+	return ln, props, func() (string, ReloadFunc) {
```
Comment (armon) [inline command/server/listener.go]: I wouldn't inline the function like this, its kind of weird to read

File: command/server.go [hunk 3]
```diff
@@ -530,6 +549,46 @@ func (c *ServerCommand) setupTelementry(config *server.Config) error {
 	return nil
 }
 
+func (c *ServerCommand) Reload(configPath []string) error {
```
Comment (vishalnayak) [inline command/server.go]: Do we need to expose this method?

Also, is it possible to use this method to load the configuration the first time as well?

File: command/server.go [hunk 4]
```diff
@@ -530,6 +548,46 @@ func (c *ServerCommand) setupTelementry(config *server.Config) error {
...
+	if config == nil {
+		retErr := fmt.Errorf("No configuration files found")
+		c.Ui.Error(retErr.Error())
+		return retErr
+	}
+
+	var reloadErrors *multierror.Error
+	// Call reload on the listeners. This will call each listener with each
+	// config block, but they verify the ID.
+	for _, lnConfig := range config.Listeners {
```
Comment (armon) [inline command/server.go]: Do we need to range over the listeners? It seems like we have one reloadFunc per anyways, seems like we can just loop over those

File: command/server/listener.go [hunk 5]
```diff
@@ -73,15 +77,52 @@ func listenerWrapTLS(
...
-	return ln, props, nil
+	return ln, props, cg.reload, nil
+}
+
+type certificateGetter struct {
+	sync.RWMutex
+
+	cert *tls.Certificate
+
+	id string
```
Comment (armon) [inline command/server/listener.go]: I don't think we need the ID

Comment (armon) [inline command/server/listener.go]: NVM, I get it.

**Comment by armon** [comment]
Minor comment, but LGTM

**Pull request merged**